src/persist-client/src/internal/machine.rs

```diff
@@ -262,31 +413,67 @@ where
                         routine,
                         compaction: compact_reqs,
                     };
-                    return Ok(Ok(Ok((seqno, writer_maintenance))));
+                    return Ok(Ok((seqno, writer_maintenance)));
                 }
-                Err(Ok(_current_upper)) => {
-                    // If the state machine thinks that the shard upper is not
-                    // far enough along, it could be because the caller of this
-                    // method has found out that it advanced via some some
-                    // side-channel that didn't update our local cache of the
-                    // machine state. So, fetch the latest state and try again
-                    // if we indeed get something different.
-                    self.fetch_and_update_state().await;
-                    let current_upper = self.upper();
-
-                    // We tried to to a compare_and_append with the wrong
-                    // expected upper, that won't work.
-                    if current_upper != batch.desc.lower() {
-                        return Ok(Ok(Err(Upper(current_upper.clone()))));
+                Err(CompareAndAppendBreak::AlreadyCommitted) => {
+                    // A previous iteration through this loop got an
+                    // Indeterminate error but was successful. Sanity check this
+                    // and pass along the good news.
```

Comment:
worth checking that `indeterminate.is_some()` here? maybe as a debug_assert and/or metric to see how often we hit this case